Mathlib/NumberTheory/Cyclotomic/Three.lean

```diff
@@ -0,0 +1,108 @@
+/-
+Copyright (c) 2024 Riccardo Brasca. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Riccardo Brasca, Pietro Monticone
+-/
+
+import Mathlib.NumberTheory.Cyclotomic.Embeddings
+import Mathlib.NumberTheory.Cyclotomic.Rat
+import Mathlib.NumberTheory.NumberField.Units.DirichletTheorem
+
+/-!
+# Third Cyclotomic Field
+We gather various results about the third cyclotomic field. The following notations are used in this
+file: `K` is a number field such that `IsCyclotomicExtension {3} ℚ K`, `ζ` is any primitive `3`-rd
+root of unity in `K`, `η` is the element in the units of the ring of integers corresponding to `ζ`
+and `λ = η - 1`.
+
+## Main results
+* `IsCyclotomicExtension.Rat.Three.Units.mem`: Given a unit `u : (𝓞 K)ˣ`, we have that
+`u ∈ {1, -1, η, -η, η^2, -η^2}`.
+
+* `IsCyclotomicExtension.Rat.Three.eq_one_or_neg_one_of_unit_of_congruent`: Given a unit
+`u : (𝓞 K)ˣ`, if `u` is congruent to an integer modulo `3`, then `u = 1` or `u = -1`.
+
+This is a special case of the so-called *Kummer's lemma* (see for example [washington_cyclotomic],
+Theorem 5.36
+-/
+
+open NumberField Units InfinitePlace nonZeroDivisors Polynomial
+
+namespace IsCyclotomicExtension.Rat.Three
+
+variable {K : Type*} [Field K] [NumberField K] [IsCyclotomicExtension {3} ℚ K]
+variable {ζ : K} (hζ : IsPrimitiveRoot ζ ↑(3 : ℕ+)) (u : (𝓞 K)ˣ)
+local notation3 "η" => (IsPrimitiveRoot.isUnit (hζ.toInteger_isPrimitiveRoot) (by decide)).unit
+local notation3 "λ" => (η : 𝓞 K) - 1
+
+/-- Let `u` be a unit in `(𝓞 K)ˣ`, then `u ∈ [1, -1, η, -η, η^2, -η^2]`. -/
+theorem Units.mem : u ∈ [1, -1, η, -η, η ^ 2, -η ^ 2] := by
+  have hrank : rank K = 0 := by
+    dsimp only [rank]
+    rw [card_eq_nrRealPlaces_add_nrComplexPlaces, nrRealPlaces_eq_zero (n := 3) K (by decide),
+      zero_add, nrComplexPlaces_eq_totient_div_two (n := 3)]
+    rfl
+  obtain ⟨⟨x, e⟩, hxu, -⟩ := exist_unique_eq_mul_prod _ u
+  replace hxu : u = x := by
+    rw [← mul_one x.1, hxu]
+    apply congr_arg
+    rw [← Finset.prod_empty]
+    congr
+    rw [Finset.univ_eq_empty_iff, hrank]
+    infer_instance
+  obtain ⟨n, hnpos, hn⟩ := isOfFinOrder_iff_pow_eq_one.1 <| (CommGroup.mem_torsion _ _).1 x.2
+  replace hn : (↑u : K) ^ ((⟨n, hnpos⟩ : ℕ+) : ℕ) = 1 := by
+    rw [← map_pow]
+    convert map_one (algebraMap (𝓞 K) K)
+    rw_mod_cast [hxu, hn]
+    simp
+  obtain ⟨r, hr3, hru⟩ := hζ.exists_pow_or_neg_mul_pow_of_isOfFinOrder (by decide)
+    (isOfFinOrder_iff_pow_eq_one.2 ⟨n, hnpos, hn⟩)
+  replace hr : r ∈ Finset.Ico 0 3 := Finset.mem_Ico.2 ⟨by simp, hr3⟩
+  replace hru : ↑u = η ^ r ∨ ↑u = -η ^ r := by
+    rcases hru with (h | h)
+    · left; ext; exact h
+    · right; ext; exact h
+  fin_cases hr <;> rcases hru with (h | h) <;> simp [h]
+
+/-- We have that `λ ^ 2 = -3 * η`. -/
+lemma lambda_sq : λ ^ 2 = -3 * η := by
```

Comment:
This name refers to local notation. Should the notation be made scoped? Or should it be an abbrev (of course then it will have parameters which hamper read/writability...)